spec.html

```diff
@@ -7636,11 +7645,15 @@ <h1>HostEnqueuePromiseJob ( _job_, _realm_ )</h1>
       <p>HostEnqueuePromiseJob is a host-defined abstract operation that schedules the Job Abstract Closure _job_ to be performed, at some future time. The Abstract Closures used with this algorithm are intended to be related to the handling of Promises, or otherwise, to be scheduled with equal priority to Promise handling operations.</p>
       <p>The _realm_ parameter is passed through to hosts with no normative requirements; it is either *null* or a Realm.</p>
 
+      <p>The implementation of HostEnqueuePromiseJob must conform to the requirements in <emu-xref href="#sec-jobs"></emu-xref>. Additionally, Jobs enqueued by this hook must conform to all of the following requirements:</p>
+      <ul>
+        <li>Let _scriptOrModule_ be GetActiveScriptOrModule(). If _realm_ is not *null*, when the Job runs, the implementation must perform implementation-defined steps such that execution is prepared to evaluate code in _scriptOrModule_ before the _job_ Abstract Closure is called.</li>
```

Comment:
I believe `_realm_` is separate from the realm we need to worry about here. The current realm record when HostEnqueuePromiseJob is called needs to be the current realm record when the job is executed.

Reply:
> The current realm record when HostEnqueuePromiseJob is called needs to be the current realm record when the job is executed.

There is no normative requirement for this.

Reply:
The job uses the current realm to create objects that are exposed to user code. How is that not a normative requirement?

Reply:
You're referring specifically to PromiseResolveThenableJob's use of CreateResolvingFunctions, yes? 

There is a normative requirement that the resolving functions are created in _some_ Realm, but not in the same Realm. I'd want the requirements on hosts to be as lax as we can make it without breaking things. Does creating the resolving functions in a different realm break anything?

(I do see that there is an ambiguity in the current PR that won't prepare to evaluate code if the thenable is a revoked proxy where _realm_ is null. That case still needs to create the resolving functions, even though they become unobservable in that case. If you're operating on the assumption that all steps are observable all the time, the [[Realm]] assert in CreateBuiltinFunction will fail.)

Reply:
The specification, before the change that underspecified this, was very clear about using the current realm. We shouldn't change that, and I see zero benefit to saying "hosts can create objects in whatever random realm they want"

Reply:
Before the promise layering change, this was dead code that was unused and also had no basis in reality given HTML's willful violation.

There is a benefit to giving hosts latitude: looser coupling and easier future layering.

I've explained several times now to you the corner case that's made easier by giving hosts this latitude about Realms for scheduled jobs: revoked proxies. The resolving functions are unobservable in that case.

Your hardline response here and in previous discussion is not productive and takes more of my time than I should like. Thanks for your input, but I'm not planning to change this part of the PR.

Reply:
I'm talking about when it *is* observable. It sounds like you just need to qualify that when a realm is required, it has to be the active realm, and when a realm isn't needed (like with your revoked proxies), t doesn't matter.

Reply:
Actually with a revoked proxy, the host could handle the exception, for example node's uncaughtException handler, which would make the exception visible to user code right?

Reply:
Good point on the TypeError that gets created in the PromiseThenableJob case; this PR needs to say PromiseThenableJobs need a valid Realm at all times.

Reply:
I think this PR is a strict improvement on the current spec as-is.

It is not clear to me if it makes sense to normatively require that the Realm of the top of the execution context stack when the job executes is the same as the Realm of the top of the execution context stack when the Host is invoked. This part of the interaction between specs/hosts/implementations is complicated enough that it is hard for me to be sure whether or not anyone actually wants to violate that potential normative requirement. It may have been required prior to the layering PR, but the point of the layering PR was that that code was not usable for most hosts, so I don't really care what it said. 

So, for now, I would propose that we add a NOTE to the line this conversation is attached to which reads

> Note: typically this results in the execution context stack having as its topmost item an execution context whose Realm component was the Realm component of the topmost item on the execution stack when HostEnqueuePromiseJob was invoked, although this is not a normative requirement

and amend that later if we become confident it should be a normative requirement.

Reply:
In the case of that TypeError, it would at least matter for V8's Error.prepareStackTrace API, since it is scoped by realm.

Reply:
Whatever V8 does is definitely not normative. We do need to say that _a_ Realm must be present, but that it must be the Realm of the active script at the time of queuing is undesirable for me, see https://github.com/tc39/ecma262/pull/1984#issuecomment-625604211.